### ISSUE TITLE
Add "type" to textDocument/reference results (icon, filtering)

### ISSUE THREAD

**Issue by sean-mcmanus** [microsoft/vscode#74237]
The C/C++ extension wants to be able to send back references of different types that users can determine what "type" of reference it is and filter out unwanted types -- in order to provide functionality equivalent to Visual Studio 2019. The types are

1. Confirmed reference (i.e. passes semantic checking): reference icon
2. Not a reference (i.e. has the same name, but fails semantic checking): cancel icon
3. Reference in a comment: comment icon
4. Reference in a string: string icon
5. Reference in an inactive (preprocessor) region: ? icon (or a better one, since it conflicts with 7)
6. Confirmation in progress: progress icon?
7. Cannot confirm reference (due to some bug or "by design" limitation: ? icon

This is the screenshot from VS:

![image](https://user-images.githubusercontent.com/19859882/58290846-02e6cc80-7d70-11e9-817a-073adcc50bfb.png)

Our current "workaround" is to only send back Confirmed References (number 1), although we also plan to have some progress UI with a button that allows users to view the Confirmation in progress (number 6 items, merged in with number 1 items).

The relevant LSP that would need to be changed is at
https://microsoft.github.io/language-server-protocol/specification#textDocument_references

**Comment by dbaeumer** [comment]
@sean-mcmanus just to clarify: you are asking for a change of the LSP request and an implementation of the UI in VS Code?

**Comment by sean-mcmanus** [comment]
@dbaeumer Yes -- I wasn't sure if I should file a separate issue on the LSP site or not.

**Comment by dbaeumer** [comment]
@sean-mcmanus yes, please file a separate issue on the LSP side since this might be tackled independently from VS Code.

**Comment by jrieken** [comment]
> Not a reference (i.e. has the same name, but fails semantic checking): cancel icon

This is funny. Searching for references returns references which are not references? 

> Confirmation in progress: progress icon?

How should that work? Will the result update when confirmation is done or does the user re-run reference search? The former doesn't fit nice into our provide-and-forget model, the latter isn't IMO the best UX

**Comment by sean-mcmanus** [comment]
Yes -- the "Not a reference" means the name is the same, but the IntelliSense type check failed. There are scenarios where users may want to still see what non-references got filtered out.

Our pending implementation for C++ Find All References is able to quickly get a list of "unconfirmed references" that match lexically, but semantic checking requires that every file with a reference needs to get compiled by our IntelliSense process, which could take a long time, so due to the inability of VS Code to allow streaming references (https://github.com/Microsoft/vscode/issues/20010), we're planning to show a popup progress dialog that shows the confirmation progress with buttons to allow users to show the "confirmed references" and/or "confirmation in progress" references (i.e. "unconfirmed" references), while we continue to continue to confirm the references until it's canceled. The user is not expected to have to re-run the references command, because our plan is to re-run it ourselves -- to enable us to return the more complete results later on. If you have a better UX design, please let us know. I think we had an email thread on this streaming references issue a few months ago.

**Comment by sean-mcmanus** [comment]
We also want a similar "type" to be usable in rename to allow users to filter out the rename results by type (I could file a separate issue if you want), but we don't plan to add Rename till July:
![image](https://user-images.githubusercontent.com/19859882/58519598-3ad38280-8169-11e9-829c-a922776e5870.png)

**Comment by jrieken** [comment]
> users to filter out the rename results by type (I could file a separate issue if you want), but we don't plan to add Rename till July

This is definitely a different issue. The rename UX/flow is not based on confirmations/uncertain data but assumes a language service with full knowledge. What you show looks more like find and replace.

**Comment by jrieken** [comment]
> So when confirmation is 100% done (or canceled) the results will update and the progress dialog will disappear.

How do you imagine that to work? API-wise?

**Comment by sean-mcmanus** [comment]
@jrieken The API for sending the the final results back is just the response message for textDocument/references, and the progress dialog was going to be a vscode.window.withProgress, but it looks like there's no option to add buttons to the progress UI for users to indicate a "show partial results"...so that is a problem :(

We'll need to come up with some alternative way. The implementation of this might not start till a week or so.

**Comment by sean-mcmanus** [comment]
Our current (revised) plan is to use some "Find All References..." status bar item (with hover showing the progress %) and clicking on the status bar icon gives a list of usable commands to control the Find All References behavior, i.e. show current confirmed and/or pending unconfirmed references and/or cancel. Let us know if you have a better UX idea.

**Comment by sean-mcmanus** [comment]
I'm exploring using a "C/C++ References" Output window pane to show all the non-confirmed type of references. Example output:

String reference: const char *s = "myFunc";
<file>:line:column [link]

Comment reference: // fix myFunc
<file>:line:column [link]

Inactive reference: int myFunc() {} // inactive definition with Windows builds
<file>:line:column [link]

Cannot confirm reference: int i = myFunc(); // type info not available here
<file>:line:column [link]

Not a reference: int myFunc; // a var
<file>:line:column [link]

Possible reference (confirmation in progress): int myFunc()
<file>:line:column [link]

Not sure yet how to best update "possible references" to another type.

**Comment by sean-mcmanus** [comment]
For V1 we used the Output window, but we're working on V2 that adds a C++ References tree view with the other types (https://github.com/microsoft/vscode-cpptools/issues/4079). So this issue could be closed unless other languages extensions want to have a similar experience built-in, e.g. if TypeScript wanted to have comment references and have a shared UI in the same tree view.
